File: Data Structures & Algorithms/find-duplicate-integer/submission-2.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        fast, slow = 0, 0

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]

            if fast == slow:
                break

        fast = 0
        while True:
            slow = nums[slow]
            fast = nums[fast]

            if fast == slow:
                break
        
        return slow

File: Data Structures & Algorithms/two-integer-sum/submission-2.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        unordered_map<int, int> sumMap;

        // [3,4,5,6]
        // 7-4 = 3

        for (int i = 0; i < nums.size(); i++) {
            int diff = target - nums[i];

            if (sumMap.count(diff) > 0) {
                // vector<int> ans(2, 0) = {sumMap[diff], i}
                return {sumMap[diff], i};
            } else {
                sumMap[nums[i]] = i;
            }
        }

        return {-1, -1};
    }
};
